enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,253 @@
+    * MCD calls nmstatectl set --no-commit /tmp/uuid-of-the-desiredState.yaml
+    * If no connectivity loss, MCD calls nmstatectl commit [PREVIOUS CHECKPOINT]
+    * Otherwise, MCD calls nmstatectl rollback
+* MCD sets the currentState to match the desiredState.
+
+#### Option B
+
+Interact with kubernetes-nmstate handler through MCD. Prepare hooks in MCD for that. User would create MCs, those would be propagated to MCD, MCD would delegate to kubernetes-nmstate handler.
+
+#### Recommendation
+
+We believe that option A poses the best combination of using existing MCO with the introduction of kubernetes-nmstate.
+
+### User Stories
+
+#### Host network configuration (Telco)
+
+As an OpenShift administrator working in Telco, I need to add and remove VLANs on day-2+, more often than the update cycle. Reboot of the whole baremetal cluster could take days and/or impact capacity.
+
+#### L2 networking for VMs (CNV)
+
+As an OpenShift administrator using Container Native Virtualization, I need to configure ad-hoc VLANs and Linux bridges on day-2. These bridges are used to provide L2 connectivity to VMs. 
+
+#### L2 networking for Pods (Telco, CNF)
+
+As an OpenShift administrator working in Telco, I need to configure bondings and ad-hoc VLANs on day-2. Macvlan ifaces will be attached to these vlans and plugged into containers.
+
+#### Declarative container native API
+
+As an OpenShift administrator using a Baremetal platform, I want to have a declarative container native API for network interface configuration rather than have to put ifcfg/NM keyfiles.
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+### Risks and Mitigations
+
+#### Not rebooting could hide errors in network configuration
+
+A NetworkManager profile is configured with certain settings (like connection.autoconnect=yes). See those settings (the content of the profiles) with `nmcli connection show "$PROFILE"`. The profile is persisted to disk (in the file you see in `nmcli -f all connection`). When NM starts, it loads the profiles from disk. NetworkManager will try to automatically connect/activate profiles that are configured to autoconnect.
+
+There is no guarantee that a reboot will succeed to connect to the network. e.g. the DHCP might simply not reply. But in general, if you have a profile (persisted to disk), and the profile is set to autoconnect, and you reboot, there is a reasonable expectation that it should work.
+
+If it doesn't come up on reboot, it's likely an environment issue (e.g. cable unplugged) or a real bug (that will be fixed by NetworkManager team).
+
+## Design Details
+
+1. User Creates **NNCP**
+2. KNMStateController takes the **NNCP** desiredConfig and puts it as a **MachineConfig** (MC) for the desired pool
```

Comment:
Does MachineConfig is proprietary to OCP? 

If the answer is Yes, I guess we can't have this code in k8s-nmstate upstream repos, right?